### ISSUE TITLE
Not working with latest update of the gateway

### ISSUE THREAD

**Issue by toblum** [nidayand/node-red-contrib-tradfri#11]
Hi,

the node is not working with the latest 1.2.42 release of the tradfri gateway. The protocol has changed. See here for reference: https://github.com/ggravlingen/pytradfri/issues/90

Regards
Tobias

**Comment by sampy14** [comment]
Same issue for me. It stopped working after the latest update

**Comment by nilsbernhar** [comment]
Same here, though i'd be surprised if the issue is with the node-red node... I would suspect the actual issue lies in the node-tradfri-argon and the coap-client...
Nonetheless, i'd be happy if someone with more skills than me fixed this ;)

**Comment by jurekbarth** [comment]
The Ikea Gateway now has "users". At the moment we connect via coaps with the following:
`coap-client -m get -u "Client_identity" -k "TRADFRI_PSK" "coaps://TRADFRI_IP:5684/15004"`

After the update you need to issue a new user and you will get back a new PSK key.
`coap-client -m post -u "Client_identity" -k "TRADFRI_PSK" -e '{"9090":"USERNAME"}' "coaps://TRADFRI_IP:5684/15011/9063"`

You will get back something like this:
```
decrypt_verify(): found 24 bytes cleartext
decrypt_verify(): found 50 bytes cleartext
{"9091":"USERNAME_PSK","9029":"1.2.0042"}
```

Afterwards you will need to connect via your new user and psk.
`coap-client -m get -u "USERNAME" -k "USERNAME_PSK" "coaps://TRADFRI_IP:5684/15004"`

@AlSa-Her You're correct, we need to change the username here:
https://github.com/nidayand/node-tradfri-argon/blob/master/lib/command-builder/index.js#L14
https://github.com/nidayand/node-tradfri-argon/blob/master/lib/command-builder/index.js#L66

**Comment by nilsbernhar** [comment]
@jurekbarth first of all, thank you very much.

Though, as I am completely new to this, I still have some questions:
1. what is the "Client_identity"
2. is "YOUR_PSK" the tradfri security id?
3. Do I have to replace the "USERNAME" or anything else (i.e. the TRADFRI_IP)?
4. Lastly, What exactly does

> we need to change the username here:

mean? (what exactly has to be changed?)

Thanks in advance!

**Comment by jurekbarth** [comment]
@AlSa-Her 

1. That's a fixed thing.
2. Yes it's the tradfri psk :)
3. You need to change everything in capital letters. 

TRADFRI_PSK --> Ikea Tradfri PSK, you can find it on the back of your Gateway.
TRADFRI_IP --> Ikea Tradfri Ip-Address, e.g. 192.168.2.142 you can find it in the router admin area.
USERNAME --> Choose something here, e.g. "mycustomuser"

4. `coap-client -m post -u "Client_identity" -k "TRADFRI_PSK" -e '{"9090":"USERNAME"}' "coaps://TRADFRI_IP:5684/15011/9063"` this command sends a coap post message with the instructions for the Ikea Gateway to create a new user. In response you will get a psk for your choosen USERNAME.

Now you need to replace "Client_identity" with USERNAME in the file linked above. And of course in your node red config change the PSK in the settings with your new USERNAME_PSK.

Note: It's a very hacky way to fix it, we need some broader solution, but it should work in the meantime. 

Sorry i'm not the best in explaining. Hope this helps anyway.

**Comment by nilsbernhar** [comment]
@jurekbarth Thanks again!

The first 3 things I figured out... I was just typing it, but your beat me to it (I originally had a typo).

The second part should be clear now. I'll try it right away!

**Comment by sampy14** [comment]
I tried this solution and I manage to connect through the terminal. However I still get an error in the Node-Red config node even though I replace the PSK with the newly created user PSK.

**Comment by nilsbernhar** [comment]
@jurekbarth everything works like a charm now!

For anyone who hasn't left node-red before this and is wondering where to type in

> `./coap-client -m post -u "Client_identity" -k "TRADFRI_PSK" -e '{"9090":"USERNAME"}' "coaps://TRADFRI_IP:5684/15011/9063"`

with the switched out "capital letter words":

go to terminal and navigate to the coap-client folder (for example libcoap/examples:
` cd /Users/USERNAME/libcoap/examples/`
replacing USERNAME with your username)

@jurekbarth 

> It's a very hacky way to fix it, we need some broader solution, but it should work in the meantime.

obviously a integration directly into the node-tradfri-argon library would be nice ;)

> Sorry i'm not the best in explaining. Hope this helps anyway.

I think you did a fairly nice job, and as long as you are willing to elaborate upon further questioning without any condescending comments, don't worry about it!

**Comment by nilsbernhar** [comment]
@sampy14 did you change the "Client_identity" in the files specified by jurekbarth in addition to changing the "Security code" in the Node-Red tradfri Node from the Security code to the newly created USERNAME_PSK?

so, replacing "Client_identity" with your newly created "USERNAME_PSK" in file (on lines 14 and 66):
`/Users/USERNAME/.node-red/node_modules/node-tradfri-argon/lib/command-builder/index.js`

**Comment by jurekbarth** [comment]
@sampy14 as @AlSa-Her said, you need to change the node-tradfri-argon command builder and replace "Client_identity" with "USERNAME" (the username you've choosen). Afterwards restart your nodered process.

**Comment by sampy14** [comment]
@AlSa-Her @jurekbarth Yes, I did replace the "Client_identity" with "USERNAME" in the index.js, as well as restarting the node-red server. Still no luck.

**Comment by jurekbarth** [comment]
@sampy14 you may issue another user and try again?

**Comment by sampy14** [comment]
Just tried issuing another user.

here is my terminal response

pi@raspberrypi:~/.node-red/node_modules/node-tradfri-argon/lib $ coap-client -m get -u "tradfri" -k "itchgMtAWUvaDstw" "coaps://10.0.0.102:5684/15004"
v:1 t:CON c:GET i:364a {} [ ]
decrypt_verify(): found 24 bytes cleartext
decrypt_verify(): found 26 bytes cleartext
[170491,178562]

It seems like the user is able to connect.

now my index.js looks like this
```
'use strict';

const isUndefined = require('lodash/isUndefined');

class CoapCommandBuilder {
    constructor(coapClientPath, securityId, hubIpAddress) {
        this.coapClientPath = coapClientPath;
        this.securityId = securityId;
        this.ip = hubIpAddress;
    }

    get(type, id) {
        const endpoint = type === 'device' ? 15001 : 15004;
	let command = `${this.coapClientPath} -m get -u "tradfri" -k "${this.securityId}" "coaps://${this.ip}:5684/${endpoint}`;


        if (id) {
            command += `/${id}"`;
        } else {
            command += '"';
        }

        return command;
    }

    put(type, id, operation) {
        const endpoint = type === 'device' ? 15001 : 15004;

        const modifier = {};
        if (!isUndefined(operation.state)) {
            if (operation.state === 'on' || operation.state === 1 || operation.state === true) {
                modifier[5850] = 1;
            } else {
                modifier[5850] = 0;
            }
        }

        if (!isUndefined(operation.transitionTime)) {
            modifier[5712] = operation.transitionTime;
        }

        if (type === 'device' && !isUndefined(operation.color)) {
            let color = operation.color.toLowerCase();
            switch (color) {
            case 'focus':    case 'cool':   color = 'f5faf6'; break;
            case 'everyday': case 'normal': color = 'f1e0b5'; break;
            case 'relax':    case 'warm':   color = 'efd275'; break;
            default:
                if ( ! color.match('/^[0-9a-f]{6}$/'))
                    color = undefined;
            }

            if (color) modifier[5706] = color;
        }

        if (!isUndefined(operation.brightness)) {
            modifier[5851] = operation.brightness;
        }

        let e = '';
        if (type === 'device') {
            e = `'{"3311":[${JSON.stringify(modifier)}]}'`;
        } else {
            e = `'${JSON.stringify(modifier)}'`;
        }

        const command = `${this.coapClientPath} -m put -u "tradfri" -k "${this.securityId}" -e ${e} "coaps://${this.ip}:5684/${endpoint}/${id}"`;

        return command;
    }

    static create(coapClientPath, securityId, hubIpAddress) {
        return new CoapCommandBuilder(coapClientPath, securityId, hubIpAddress);
    }
}

module.exports = CoapCommandBuilder;

```

And I have input the new PSK in the Tradfri node config.

Still it returns an Hub config error

**Comment by jurekbarth** [comment]
@sampy14 everything looks ok to me. Your username/psk combination is fine. You can try to reinstall the node-red-contrib-tradfri package and change the code again and restart again. I'm running my installation in a docker container, but i can try to set up an raspberry pi later today and see if i can get it up and running.

**Comment by sampy14** [comment]
@jurekbarth A reinstall of both the node-red-contrib-tradfri and the lib didn't help. It is very strange as I can send commands to the gateway without issues.

**Comment by thunder-62** [comment]
@sampy14 just for debugging: I would create a short shell script "echo $@ > /tmp/test" and put it in the coap-path instead of <path/>coap-client. so you can verify that the username "Client_identity" is really replaced by your username.

**Comment by jurekbarth** [comment]
@sampy14 you may try this updated version.

1. Uninstall node-red-contrib-tradfri with `npm uninstall -S node-red-contrib-tradfri`
2. Install node-red-contrib-tradfri from github with: `npm install --save https://github.com/jurekbarth/node-red-contrib-tradfri`
3. Update your node-red-config

Hopefully this works. It could be even better if we could issue a new user from the nodered admin panel.

**Comment by sampy14** [comment]
@jurekbarth thanks for the support! 

After a fresh install of node red your version of the tradfri node worked perfectly.

I agree that it could be great to be able generate a new user from the config panel.

**Comment by THM11** [comment]
@sampy14 @AlSa-Her 
I'm really like fish out of water when it comes to "out-side node-red" (aka terminal). I tried to follow the steps above, but if I try any of the commands mentioned above, I get some bash error...

pi@raspberrypi:~/.node-red/node_modules/node-tradfri-argon/lib $ coap-client -m get -u "Client_idenity" -k "SecurityCodeHere" "coaps://192.168.1.205:5684/15004"
bash: coap-client: command not found

Basically, would it be in anyway possible to get even more "step-by-step" instructions on which folder to actually go (assuming username of the Raspi is pi) etc.. :)

**Comment by toblum** [comment]
This is clearly something the has to be fixed in the underlying node-tradfri-argon library, but there in no possibility to file issues. I will give your modified version a try soon.

I saw that the author of the original node-tradfri library (from wich the node-trafri-argon library was forked from) wrote that he will release a updated version soon: https://github.com/morzzz007/node-tradfri/issues/7

**Comment by donster22** [comment]
@THM11i'm having exactly the same problem as you with: bash: coap-client: command not found
....and I thought I was reasonably competent on the command line. i hadn't even considered an update could cause this problem and with all of my new lights automated by node red i'm a bit stuffed unless someone can suggest how i could run that coap-client command please? thanks.

**Comment by nilsbernhar** [comment]
@THM11 and @donster22 sorry for the confusion!

My mistake... coap-client is a file.
To execute files in the current directory, you have to put a ./ in front of the file name! I'm terribly sorry I forgot that...

I'll change it in the original post, but the actual line should be:

`./coap-client -m post -u "Client_identity" -k "TRADFRI_PSK" -e '{"9090":"USERNAME"}' "coaps://TRADFRI_IP:5684/15011/9063"`

I hope this works for you.

**Comment by jurekbarth** [comment]
@THM11 and @donster22 

I've update the version of my node-red-contrib-tradfri, it's possible to generate a new user from within the node-red-webinterface. 

Very detailed for the raspberry pi:
1. `cd ~/.node-red` go to your node-red directory
2. `node-red-stop` stop the node-red service
3. `npm uninstall -S node-red-contrib-tradfri` uninstall this package
4. `npm install --save https://github.com/jurekbarth/node-red-contrib-tradfri` install new version directly from github
5. `node-red-start` starts the node-red service again
6. go to your node-red-contrib-tradfri config and use the fields below "generate new user-indentity" 
```
hubip: <your_ikea_tradfri_ip>
tradfri psk code: <your_ikea_tradfri_code>
choose a username: customtradri (or something else)
CoAP path: /home/pi/.node-red/node_modules/node-red-contrib-tradfri/node_modules/node-tradfri-argon/lib/coap-client-raspbian
```
7. click generate user-identity --> it should fill in the gaps above :)
8. Save it and you may store the username/psk 

Hopefully this works :)

**Comment by donster22** [comment]
Thanks @jurekbarth . Sorted last night once Id put the “./“ infront of the command mentioned and realised I had to change 2 entries in the index.js file rather than just the 1. Will update as per your new instructions tonight.

**Comment by muracz** [comment]
@jurekbarth for me the solution worked like a charm. Well done, thanks!

**Comment by nilsbernhar** [comment]
@jurekbarth Thanks for the updated version! It works great!

I just set up a new raspberry Pi Zero W with dietpi. Somehow, the old node-red-contrib-tradfri didn't work anymore (this could be a general problem since the node-tradfri-argon got updated 5 hours ago - I am not sure), but yours works.

Thanks for sharing.
Much praise from me, once again!

**Comment by nidayand** [comment]
Apologies. I've been unavailable (sick) and I've not been monitoring the issue log. I've been working on in but I had not committed the code. I noticed a lot have been done but it breaks a bit the dependencies to argon so I'll be going with my contrib to the master. I've just published 1.1.0 to npm. Now also supports additional git pulls such as RGB

**Comment by toblum** [comment]
For me, it works again. Thank you for updating the library and bringing new functions.

Regards
Tobias

**Comment by nidayand** [comment]
Thanks. Fyi.. I'm in the process of migrating from node-tradfri-argon to node-tradfri-client (nodejs v6 dependency). This will make the package pure a pure nodejs distribution

**Comment by orandaadnaro** [comment]
Hi - Ive just briefly read through the above posts. As I've installed the hassio version of home assistant and using todays version of node-red, I'm wondering do I have to make these changes by accessing my pi over a terminal connection vssh ... or is there a way of making these changes from within hassio?


I've a tradfri motion sensor and 3 bulbs I want to control, but I can't get the gateway to get a preshared key using a app id.
